cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/DefaultProvisionerConfigProvider.java

```diff
@@ -47,55 +47,70 @@
-          } catch (Exception e) {
-            LOG.warn("Exception reading JSON config file for provisioner {}. Ignoring file.",
-                     provisionerName, e);
-          }
-        }
+        results.put(provisionerName, GSON.fromJson(new JsonReader(reader), ProvisionerConfig.class));
+      } catch (Exception e) {
+        LOG.warn("Exception reading JSON config file for provisioner {}. Ignoring file.", provisionerName, e);
       }
     }
+
     return Collections.unmodifiableMap(results);
   }
+
+  /**
+   * Find the json config file with name [provisioner_name].json under the provisioner ext directory.
+   * The ext directory can be found by the classloader of the provisioner.
+   * If the json config file is not found in the ext directory, try to locate it from the provisioner classloader.
+   *
+   * @param provisioner the {@link Provisioner}
+   * @return a {@link Reader} for reading the config file, or {@code null} if the config file cannot be located
+   */
+  @Nullable
+  private Reader openConfigReader(Provisioner provisioner) throws IOException, URISyntaxException {
+    Class<? extends Provisioner> provisionerClass = provisioner.getClass();
+    URL classPathURL = ClassLoaders.getClassPathURL(provisionerClass);
+
+    // This shouldn't happen as we should be able to find the class from its own classloader
+    if (classPathURL == null) {
+      LOG.warn("Unable to find provisioner class {} from its own classloader", provisionerClass);
+      return null;
+    }
+
+    // Find the json config file with name [provisioner_name].json under the provisioner ext directory.
+    // The ext directory can be found by the classloader of the provisioner.
+    // See AbstractExtensionLoader class for the directory structure
+    String provisionerName = provisioner.getSpec().getName();
+    File configFileDir = new File(classPathURL.toURI());
+
+    // If the class is loaded from a (jar) file,
+    // then the .json config file is expected to be at the parent directory of it.
+    if (Files.isRegularFile(configFileDir.toPath())) {
+      configFileDir = configFileDir.getParentFile();
+    }
+
+    Path configFilePath = new File(configFileDir, provisionerName + ".json").toPath();
+    if (Files.isRegularFile(configFilePath)) {
+      return Files.newBufferedReader(configFilePath, StandardCharsets.UTF_8);
+    }
+
+    // If we cannot find the config file in the filesystem, try to look for the json config from the classloader itself.
+    // This is the case when the .json is packaged inside the jar (e.g. system provisioners).
+    URL resource = provisionerClass.getClassLoader().getResource(provisionerName + ".json");
+    if (resource != null) {
+      return new InputStreamReader(resource.openStream(), StandardCharsets.UTF_8);
+    }
+
+    LOG.debug("Unable to find JSON config file {} for provisioner {}", configFilePath, provisionerName);
```

Comment:
native provisioner config will always be unfindable - do we want to skip trying to load its config?

Reply:
It only showed once as a debug log, I think it's ok not to special handle the native profile in this class.